src/brod_client.erl

```diff
@@ -418,21 +418,41 @@ shutdown_pid(_) ->
 
 -spec get_partition_worker(client(), partition_worker_key()) ->
         {ok, pid()} | {error, get_worker_error()}.
-get_partition_worker(ClientPid, Key) when is_pid(ClientPid) ->
-  case erlang:process_info(ClientPid, registered_name) of
-    {registered_name, ClientId} when is_atom(ClientId) ->
-      get_partition_worker(ClientId, Key);
+get_partition_worker(Client, Key) when is_atom(Client) ->
+  case erlang:whereis(Client) of
+    ClientPid when is_pid(ClientPid) ->
+      case erlang:is_process_alive(ClientPid) of
+        true -> lookup_partition_worker(Client, ?ETS(Client), Key);
+
+        _ ->
+          %% This is an attempt to help with a race condition
+          %% between ets lookups and what is being supervised.
+          %% Adding a gratuitous call to `get_workers_table`
+          %% to flush the Client's inbox of any inflight worker registration or deregistration.
+          get_partition_worker_with_ets(ClientPid, Key)
+      end;
+    _ -> {error, client_down}
+  end;
+
+get_partition_worker(ClientPid, Key) when is_pid(ClientPid)->
+  case {erlang:is_process_alive(ClientPid), erlang:process_info(ClientPid, registered_name)} of
```

Comment:
No need to check liveness together with process_info, because process_info always return undefined for dead process.